pkg/sharing/files.go

```diff
@@ -349,15 +350,44 @@ func (s *Sharing) UpdateDir(inst *instance.Instance, target map[string]interface
 	}
 	revisions, ok := target["_revisions"].(map[string]interface{})
 	if !ok {
+		inst.Logger().WithField("nspace", "replicator").
+			Warnf("Missing _revisions for updating directory %#v", target)
 		return ErrInternalServerError
 	}
+	oldDoc := dir.Clone().(*vfs.DirDoc)
 	indexer := NewSharingIndexer(inst, &bulkRevs{
 		Rev:       rev,
 		Revisions: revisions,
 	})
-	oldDoc := dir.Clone().(*vfs.DirDoc)
+	fs := inst.VFS().UseSharingIndexer(indexer)
 	copyTagsAndDatesToDir(target, dir)
-	// TODO what if name or dir_id has changed
+	name, ok := target["name"].(string)
+	if !ok {
+		inst.Logger().WithField("nspace", "replicator").
+			Warnf("Missing name for updating directory %#v", target)
+		return ErrInternalServerError
+	}
+	dir.DocName = name
+	inst.Logger().WithField("nspace", "replicator").
+		Debugf("Update dir name: %s", name)
+	if dirID, ok := target["dir_id"].(string); ok {
+		if dirID != dir.DirID {
+			parent, err := fs.DirByID(dirID)
+			// TODO better handling of this conflict
+			if err != nil {
+				inst.Logger().WithField("nspace", "replicator").
+					Debugf("Conflict for parent on updating dir: %s", err)
+				return err
+			}
+			dir.DirID = parent.DocID
+		}
+	} else {
```

Comment:
In which case the `dir_id` wouldn't exist?

Reply:
When the parent directory is not the same on both instances. For example, it can be files shared by references (the files are shared, but not the directory in which they are).